bullet-featherstone: Support joint spring stiffness

iche033 opened the pull request:
# 🎉 New feature

Closes https://github.com/gazebosim/gz-physics/issues/549

## Summary

Add support for joint spring stiffness in bullet-featherstone physics plugin.

The `spring_stiffness` and `spring_reference` parameters are now parsed from SDF. 
The joint damping function (now named `AddJointDampingStiffnessTorque`) is extended to apply explicit spring stiffness torque based on the input SDF values.

This should also work for ball joints but with a minor caveat:
    * ball joints have 3 DOFs. For now I parse the `spring_stiffness` and `spring_reference` parameters from `//joint/axis/dynamics` (ignoring `//joint/axis2/dynamics`) and applies these values to all 3 DOFs of the ball joint.
 
## Test it

Added a test that verifies a model with prismatic joint and non-zero spring stiffness value will cause the link to oscillate up and down. Idea taken from gazebo-classic's [spring damper test](https://github.com/gazebosim/gazebo-classic/blob/b22c6e15e52299865b31093b8feebc9ca19e26e8/test/integration/joint_test.cc#L147)

To run the test, from your colcon workspace:

```
cd build/gz-physics
ctest -R simulation_features_bullet-featherstone -V
```

You can also test with this example world

<details><summary>spring.sdf</summary>

```xml
<?xml version="1.0" ?>
<sdf version="1.12">
  <world name="spring">
    <physics name="1ms" type="ignored">
      <max_step_size>0.002</max_step_size>
      <real_time_factor>1.0</real_time_factor>
    </physics>

    <scene>
      <ambient>1.0 1.0 1.0 1.0</ambient>
      <background>0.8 0.8 0.8 1.0</background>
      <grid>false</grid>
      <origin_visual>false</origin_visual>
    </scene>

    <light type="directional" name="sun">
      <cast_shadows>true</cast_shadows>
      <pose>0 0 10 0 0 0</pose>
      <diffuse>0.8 0.8 0.8 1</diffuse>
      <specular>0.2 0.2 0.2 1</specular>
      <direction>-0.5 0.1 -0.9</direction>
    </light>

    <model name="spring">
      <pose>0 0.0 1.0 0 0 0</pose>
      <static>false</static>
      <link name="base" />
      <link name="body">
        <inertial>
          <mass>2</mass>
        </inertial>
        <collision name="collision">
          <geometry>
            <box>
              <size>0.1 0.1 0.1</size>
            </box>
          </geometry>
        </collision>
        <visual name="visual">
          <geometry>
            <box>
              <size>0.1 0.1 0.1</size>
            </box>
          </geometry>
          <material>
            <diffuse>0.0 0.8 0.0 1</diffuse>
            <specular>0.0 0.0 0.0 1</specular>
          </material>
        </visual>

      </link>
      <joint name="world_joint" type="fixed">
        <parent>world</parent>
        <child>base</child>
      </joint>

      <joint name="test_joint" type="prismatic">
        <parent>base</parent>
        <child>body</child>
        <axis>
          <xyz>0 0 1</xyz>
          <dynamics>
            <spring_stiffness>100.0</spring_stiffness>
          </dynamics>
        </axis>
      </joint>
    </model>

    <model name="spring_revolute">
      <pose>-1.0 0 1.1 0 0.0 0.0</pose>
      <link name="base" />
      <link name="body">
        <inertial>
          <inertia>
            <ixx>0.01</ixx>
            <ixy>0.0</ixy>
            <ixz>0.0</ixz>
            <iyy>0.01</iyy>
            <iyz>0.0</iyz>
            <izz>0.01</izz>
          </inertia>
          <mass>1.0</mass>
        </inertial>
        <visual name="visual_sphere">
          <geometry>
            <sphere>
              <radius>0.1</radius>
            </sphere>
          </geometry>
          <material>
            <diffuse>1.0 0.0 0.0</diffuse>
          </material>
        </visual>
        <collision name="collision_sphere">
          <geometry>
            <sphere>
              <radius>0.1</radius>
            </sphere>
          </geometry>
        </collision>
      </link>
      <joint name="world_joint" type="fixed">
        <parent>world</parent>
        <child>base</child>
      </joint>
      <joint name="test_joint" type="revolute">
        <pose>-1.0 0.0 0 0.0 0.0 0.0</pose>
        <parent>base</parent>
        <child>body</child>
        <axis>
          <xyz>0.0 -1.0 0.0</xyz>
          <dynamics>
            <spring_stiffness>100.0</spring_stiffness>
          </dynamics>
        </axis>
      </joint>
      <static>false</static>
    </model>

  </world>
</sdf>
```
</details>

This world contains a model with link connected by a prismatic joint with spring stiffness set to `100`.
   * update: Added another model (red sphere) that has a revolute joint with spring stiffness = 100

Launch `spring.sdf` with bullet-featherstone physics plugin

```sh
gz sim -v 4 spring.sdf --physics-engine libgz-physics-bullet-featherstone-plugin
```

Play sim and you should see the link fall due gravity but it starts to oscillate up and down with a spring-like behavior

![bullet_spring_stiffness](https://github.com/user-attachments/assets/2308fab6-be74-436a-886b-039af0801188)

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers
- [ ] Was GenAI used to generate this PR? If so, make sure to add "Generated-by" to your commits. (See [this policy](https://osralliance.org/wp-content/uploads/2025/05/OSRF-Policy-on-the-Use-of-Generative-Tools-Generative-AI-in-Contributions.pdf) for more info.)

Generated-by: Remove this if GenAI was not used.

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` and `Generated-by` messages.

roncapat commented:
Do you have a test world also for a revolute joint with spring parameters? I tried this one attached also on DART and works. I tried again making one on my own for revolutes and it seems not to work, at least on DART.

iche033 commented:
> Do you have a test world also for a revolute joint with spring parameters? I tried this one attached also on DART and works. I tried again making one on my own for revolutes and it seems not to work, at least on DART.

I added another test for revolute joints in [14712a1](https://github.com/gazebosim/gz-physics/pull/805/commits/14712a1673f337a9b7f34d4fac582ff6ef81203f). I also updated the `spring.sdf` test world in the PR description above to include a model with revolute joint. I see the spring behavior in both bullet-featherstone and dart.

iche033 commented on bullet-featherstone/src/Base.cc:
```diff
@@ -76,6 +79,18 @@ void GzMultiBody::SetJointPosForDof(
 
 btScalar GzMultiBody::GetJointPosForDof(int _jointIndex, std::size_t _dof) const
 {
+  if (this->getLink(_jointIndex).m_jointType == btMultibodyLink::eSpherical)
```
updated `SetJointPosForDof` in [e822eb4](https://github.com/gazebosim/gz-physics/pull/805/commits/e822eb4f19d6737f253ebe449878c0840d23dafe)

shameekganguly commented on bullet-featherstone/src/Base.cc:
```diff
@@ -69,15 +69,38 @@ void GzMultiBody::SetJointPosForDof(
+    math::Vector3 posVec = axis * angle;
+    posVec[_dof] = _value;
+    angle = posVec.Length();
+    posVec /= angle;
```
We should guard for angle = 0 here,
```suggestion
  if (fabs(angle) < 1e-10)
  {
    positions[0] = 0;
    positions[1] = 0;
    positions[2] = 0;
    positions[3] = 1;
  } 
  else
  {
    posVec /= angle;
   ...
   }
   this->setJointPosMultiDof(_jointIndex, positions);
```

iche033 replied:
added check in [b1106e2](https://github.com/gazebosim/gz-physics/pull/805/commits/b1106e2823aa3d88a7fbd958d2f63986ab3e387b)

shameekganguly reviewed (approved):
Left one comment, but looks good otherwise!
